### ISSUE TITLE
Improve module ergonomics

### ISSUE THREAD

**Issue by hegemonic** [jsdoc/jsdoc#1645]
JSDoc users have many well-justified complaints about the way JSDoc supports modules. Those complaints tend to fall into at least one of these categories:

+ **Syntax compatibility**. JSDoc's syntax for modules is not compatible with VS Code/TypeScript/Closure Compiler/(your favorite tool here).

    For example, module longnames must start with `module:` (for example, `module:foo/bar`), which other tools don't recognize.

+ **Code-parsing smarts**. JSDoc is not as good as it could be at extracting information from your code.

    For example, suppose the module `a.js` is documented as `module:a` and exports the class `Foo`, and the module `b.js` does something like `const { Foo } = require('a')`. JSDoc knows about `module:a.Foo`, and it knows that module `b` has a variable named `Foo`, but it doesn't figure out that they both refer to the same thing.

It would be worthwhile to support some other syntax that's compatible with popular tools (most notably TypeScript and Closure Compiler) and IDEs (most notably VS Code). **We should not invent a new syntax that is incompatible with the rest of the JS/TS universe** (contra [xkcd](https://xkcd.com/927/)). Also, we should continue to support the existing module syntax unless that's absolutely not possible.

Nota bene: Adding support for a different module syntax would almost certainly require large, pervasive changes to JSDoc.

**Comment by brettz9** [comment]
+1 for TypeScript

**Comment by ernestostifano** [comment]
I agree that any eventually adopted module syntax should be the most widely compatible. Additionally, I understand that changing the current one could be problematic. However, there is a (hope minor) change that could solve half of the problems.

As mentioned in other issues (#969 and especially #1533), currently, people is forced to use "module:my/module" prefix even inside the same module where types are defined. That is weird and should be addressed independently from the module syntax argument. Please note that current behaviour may not perfectly reflect what stated in the [documentation](https://jsdoc.app/tags-module.html).

> All symbols in the file are assumed to be members of the module unless documented otherwise.

Shouldn't this logic apply to assignment statements too?

```js
/** @module myModule */

// DECLARING
/** @typedef {boolean} myType */ // -> Per specs -> module:myModule~myType

// ASSIGNING
/** @const {myType} */ // -> Shouldn't this be implicitly interpreted as -> module:myModule~myType ?
const myConst = true;
```

If that gets resolved, use cases for the current module syntax would be reduced to the case in which types need to be used across different modules.

And even in that case, additional module syntax might be not needed. Let's also remember that following modules paradigm, anything declared inside a module must remain encapsulated into that module unless explicitly exported. Currently, JSDoc is only abusing the global scope by adding longer names to things (adding the module path).

However, it is interesting to note that the following pattern currently seems to work:

`./file-a.js`
```js
/** @module moduleA */

/**
* @typedef anyJavascriptContext.myType
* @property {boolean} prop1
*/

const anyJavascriptContext = true;

export {anyJavascriptContext};
```

`./file-b.js`
```js
/** @module moduleB */

import {anyJavascriptContext} from './file-a.js'

/**
* @param {anyJavascriptContext.myType} someArg
* @returns {boolean}
*/
function someFunction(someArg) {
    return someArg.prop1;
}

export {someFunction};
```

WebStorm flawlessly understands it. TypeScript correctly generates declaration files with import/export statements for types while checking above files. ESLint JSDoc plugin does not complain. Seems like the perfect solution right? (The only problem someone could point is that a real JavaScript context is needed to "carry" the type definitions. But returning to the modules paradigm, why should you use a type from a module that you are not importing to be used in some way?). Using namespaces with this pattern works too.

LONG STORY SHORT: I think that #1533 should be fixed right away (I would be glad to help). And current module syntax for cross module types complete abandonment should be considered.

**Comment by danielo515** [comment]
So typescript import syntax is not yet supported, but is there at least a way to tell the parsers to ignore it?

**Comment by wojtek-krysiak** [comment]
@danielo515 you can use `typedef-imports` plugin from [better-docs](https://github.com/SoftwareBrothers/better-docs#typedefimport) but it will have to be one liner:

```
/** @typedef {import('./some-other-file').ExportedType} ExportedType */
```

**Comment by CoachRDeveloper** [comment]
+1 for these type of type definitions.
```@returns {import("../Config/Typedefs").TextInfo}``` now gives a JSDoc error. But this is the only way VS Code will have intellisense...

**Comment by BlackWild** [comment]
While this issue is being resolved, is there a way to make JSDoc use TypeScript parser for document generation? My project is in pure JS and I don't use TypeScript but VSCode completely recognizes following syntax for JSDoc types in `.js` files and I am enjoying in-editor documentations thanks to this feature: 
- `{typeof someType}`,
- `{import('./someFile').myType}`

however, JSDoc itself does not generate documentation from this syntax and throws errors when I call `yarn jsdoc`.

Since VSCode understands this, I guess it should be a way to direct JSDoc to use any tool that VSCode is using to generate the actual documentation!

**Comment by CoachRDeveloper** [comment]
Beste jsdoc/jsdoc,

Wij hebben je e-mail ontvangen en geregistreerd onder ticket nummer #101829960.

Binnen 1 werkdag ontvang je een reactie.

Voor een snellere reactie kan je ons ook bereiken via WhatsApp op nummer +31 85 001 3385.Heb je een urgent probleem, dan zijn wij tijdens kantooruren telefonisch bereikbaar op +31 85 001 3385.



Met vriendelijke groet,

Team VIVITION

__________________________________________________



Pascalstraat 18H
6716AZ Ede
The Netherlands

T: +31(0) 85 001 3385

Website: www.VIVITION.com
Email: support@VIVITION.com
__________________________________________________

**Comment by brettz9** [comment]
@BlackWild : I am not a regular user of it, but I believe you may find tools that work against TypeScript, since it is [generally](https://stackoverflow.com/questions/29918324/is-typescript-really-a-superset-of-javascript) a strict superset of JavaScript. The preeminent tool is [typedoc](https://typedoc.org/) and it can apparently be run against JavaScript only; see [the discussion beginning with this comment](https://github.com/TypeStrong/typedoc/issues/643#issuecomment-353851978) of an issue.

One other interesting project is https://tsdoc.org/ which is Microsoft's tool to standardize AST for TypeScript comment blocks. They have a playground at https://microsoft.github.io/tsdoc/ which even generates a sample page of documentation, though I don't see mention anywhere of any published tool allowing static documentation file building, but perhaps it may be worth watching the parent tsdoc project to see if they add one; it currently seems limited to parsing or linting, however.

(As an aside on linting, FWIW, though it does apparently allow for at least allowing arbitrary tags: https://github.com/microsoft/tsdoc/tree/master/tsdoc-config , I don't know how well tsdoc meets the needs of plain JavaScript--for example, the playground seems to parse `@param` types as "nonstandard" (given that this is for TS after all where one can just put types in the param signature, but that is a bit concerning from a JS-only perspective). Their ESLint plugin currently does not seem to allow for any rule customization like our `eslint-plugin-jsdoc`. And despite its very promising development of a standard for doc block AST, its eslint plugin unfortunately does not offer a hybrid tsdoc/typescript-eslint parser which would allow for easier design of custom ESLint rules, but a third party might be able to splice these together. In any case, it does seem like a very interesting project to watch.)

**Comment by CoachRDeveloper** [comment]
Beste jsdoc/jsdoc,

Wij hebben je e-mail ontvangen en geregistreerd onder ticket nummer #101993155.

Binnen 1 werkdag ontvang je een reactie.

Voor een snellere reactie kan je ons ook bereiken via WhatsApp op nummer +31 85 001 3385.Heb je een urgent probleem, dan zijn wij tijdens kantooruren telefonisch bereikbaar op +31 85 001 3385.



Met vriendelijke groet,

Team VIVITION

__________________________________________________



Pascalstraat 18H
6716AZ Ede
The Netherlands

T: +31(0) 85 001 3385

Website: www.VIVITION.com
Email: support@VIVITION.com
__________________________________________________

**Comment by nbilyk** [comment]
While waiting on this, I made a small jsdoc plugin that solves the simple case of allowing typescript-style imports when running jsdoc.

It will convert typescript doc imports:
```js
/** @typedef {import('./model').Foo} Bar */
```

Into the jsdoc module import:
```js
/** @typedef {module:model~Foo} Bar */
```

Give it a try here:
[https://github.com/polyforest/jsdoc-tsimport-plugin/](https://github.com/polyforest/jsdoc-tsimport-plugin/)

**Comment by aaclayton** [comment]
I think that the functionality that @hegemonic mentions in <https://github.com/jsdoc/jsdoc/issues/1645#issuecomment-609999618> is a really important improvement for JSDoc to receive. Developers should be able to write a simple module like:

```js
/**
 * @module example
 */

/**
 * @typedef {Object} ModuleOptions
 * @property {string} foo
 * @property {number} bar
 */

/**
 * Some module level class
 * @param {ModuleOptions} [options]
 */
class ModuleClass {
  constructor(options={}) {
    // ...
  }
}
```

This seems like a **really basic** setup that should absolutely be supported out of the box. Both the `ModuleOptions` typedef and the `ModuleClass` class definition should be appropriately scoped `module:example` . 

IDEs like WebStorm correctly interpret this documentation, but JSDoc does not build the correct documents in this case. It is possible to correct the built documents by explicitly including `module:example` as a prefix for every reference but this is (A) not ergonomic for the developer and (B) breaks inspections inside the IDE.

I don't have enough familiarity with what is under-the-hood in JSDoc to propose how to solve this, but what my example above illustrates is (in my opinion) the really bare bones of ES Modules which are currently poorly supported in what should be the de-facto standard tool for documenting modern JavaScript (JSDoc).

**Comment by thoughtsunificator** [comment]
There is no need for a new syntax because there already is one:

https://jsdoc.app/tags-module
https://jsdoc.app/howto-es2015-modules

The real mystery here being why is the standard syntax being ignored by vendors.
